src/tp_engine.hpp

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+  void SetSpareNodeCount(const size_t node_spare_count) {
+    node_spare_count_ = node_spare_count;
+    if (using_likelihoods_) {
+      pv_likelihoods_.SetSpareNodeCount(node_spare_count);
+    }
+    if (using_parsimony_) {
+      pv_parsimony_.SetSpareNodeCount(node_spare_count);
+    }
+  }
+  void SetAllocatedNodeCount(const size_t node_alloc) {
+    node_alloc_ = node_alloc;
+    if (using_likelihoods_) {
+      pv_likelihoods_.SetAllocatedNodeCount(node_alloc);
+    }
+    if (using_parsimony_) {
+      pv_parsimony_.SetAllocatedNodeCount(node_alloc);
+    }
+  }
+
+  // Edge Counts
+  size_t GetEdgeCount() const { return edge_count_; };
+  size_t GetSpareEdgeCount() const { return edge_spare_count_; };
+  size_t GetAllocatedEdgeCount() const { return edge_alloc_; };
+  size_t GetPaddedEdgeCount() const { return GetEdgeCount() + GetSpareEdgeCount(); };
+  size_t GetSpareEdgeIndex(const size_t edge_offset) const {
+    const size_t edge_scratch_size = GetPaddedEdgeCount() - GetEdgeCount();
+    Assert(edge_offset < edge_scratch_size,
+           "Requested edge_offset outside of allocated scratch space.");
+    return edge_offset + GetEdgeCount();
+  }
+
+  void SetEdgeCount(const size_t edge_count) { edge_count_ = edge_count; }
+  void SetSpareEdgeCount(const size_t edge_spare_count) {
+    edge_spare_count_ = edge_spare_count;
+  }
+  void SetAllocatedEdgeCount(const size_t edge_alloc) { edge_alloc_ = edge_alloc; }
+
+  // ** Access
+
+  NucleotidePLVRefVector &GetLikelihoodPVs() { return pv_likelihoods_.GetPVs(); }
+  NucleotidePLVRefVector &GetParsimonyPVs() { return pv_parsimony_.GetPVs(); }
+  EigenVectorXd &GetBranchLengths() { return branch_lengths_; }
+
+ protected:
+  // ** DAG
+  // Un-owned reference DAG.
+  GPDAG &dag_;
+  // For adding temporary NNIs to DAG.
+  std::unique_ptr<GraftDAG> graft_dag_;
+
+  // ** Parameters
+  // Total number of nodes in DAG. Determines sizes of data vectors indexed on nodes.
+  size_t node_count_ = 0;
+  size_t node_alloc_ = 0;
+  size_t node_spare_count_ = 2;
+  // Total number of edges in DAG. Determines sizes of data vectors indexed on edges.
+  size_t edge_count_ = 0;
+  size_t edge_alloc_ = 0;
+  size_t edge_spare_count_ = 3;
+  // Growth factor when reallocating data.
+  constexpr static double resizing_factor_ = 2.0;
+
+  // Branch length parameters for DAG.
+  EigenVectorXd branch_lengths_;
+  // Initial branch length during first branch length opimization.
+  static constexpr double default_branch_length_ = 0.1;
+  // Absolute lower bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double min_log_branch_length_ = -13.9;
+  // Absolute upper bound for possible branch lengths during optimization (in log
+  // space).
+  static constexpr double max_log_branch_length_ = 1.1;
+
+  // ** Scoring
+  // Partial Vector for storing Likelihood scores.
+  bool using_likelihoods_;
+  PSVHandler pv_likelihoods_;
```

Comment:
Do you think `likelihood_pvs` is more appropriate? (ditto below)

Reply:
Yes. Name changed.